File: DemoServer/AdditionalLanguages/python/multiMapIndexes/multiMapIndexCustomizedFields/multiMapIndexCustomizedFields.py
from demo_example import Example, RunParamsBase

#region Usings
from ravendb.documents.indexes.index_creation import AbstractMultiMapIndexCreationTask
from ravendb.documents.indexes.definitions import FieldStorage
#endregion

class RunParams(RunParamsBase):
    def __init__(self, name_prefix: str = None, title_prefix: str = None):
        self.name_prefix = name_prefix
        self.title_prefix = title_prefix

#region Demo
#region Step_1
class Contacts_ByNameAndTitle(AbstractMultiMapIndexCreationTask):
#endregion
    #region Step_2
    class IndexEntry:
        def __init__(
            self,
            contact_name: str = None,
            contact_title: str = None,
            collection: object = None,
        ):
            self.contact_name = contact_name
            self.contact_title = contact_title
            self.collection = collection
    #endregion

    #region Step_3
    class ProjectedEntry(IndexEntry):
        def __init__(
            self,
            contact_name: str = None,
            contact_title: str = None,
            collection: object = None,
            phone: str = None,
        ):
            super().__init__(contact_name, contact_title, collection)
            self.phone = phone
    #endregion

    def __init__(self):
        #region Step_4
        super().__init__()
        self._add_map(
            (
                "docs.Employees.Select(employee => new {"
                '    contact_name = (employee.FirstName + " ") + employee.LastName,'
                "    contact_title = employee.Title,"
                '    collection = this.MetadataFor(employee)["@collection"]'
                "})"
            )
        )

        self._add_map(
            (
                "docs.Companies.Select(company => new {"
                "    contact_name = company.Contact.Name,"
                "    contact_title = company.Contact.Title,"
                '    collection = this.MetadataFor(company)["@collection"]'
                "})"
            )
        )

        self._add_map(
            (
                "docs.Suppliers.Select(supplier => new {"
                "    contact_name = supplier.Contact.Name,"
                "    contact_title = supplier.Contact.Title,"
                '    collection = this.MetadataFor(supplier)["@collection"]'
                "})"
            )
        )
        #endregion
        #region Step_5
        self._store("contact_name", FieldStorage.YES)
        self._store("contact_title", FieldStorage.YES)
        self._store("collection", FieldStorage.YES)
        #endregion
#endregion

class MultiMapReduceIndexCustomizedFields(Example):
    def run(self, run_params: RunParams):
        namePrefix = run_params.name_prefix or "Michael"
        titlePrefix = run_params.title_prefix or "Sales"
        Contacts_ByNameAndTitle().execute(self.document_store_holder.store())
        #region Demo        
        #region Step_6
        with self.document_store_holder.store().open_session() as session:
            contacts = list(
                session.query_index_type(Contacts_ByNameAndTitle, Contacts_ByNameAndTitle.IndexEntry)
                .where_starts_with("contact_name", namePrefix)
                .where_starts_with("contact_title", titlePrefix)
                .select_fields(Contacts_ByNameAndTitle.ProjectedEntry)
            )
        #endregion
        #endregion

        return contacts

File: DemoServer/AdditionalLanguages/php/multiMapIndexes/multiMapIndexCustomizedFields/MultiMapIndexCustomizedFields.php
<?php

namespace RavenDB\Demo\multiMapIndex\multiMapIndexCustomizedFields;

//region Usings
use RavenDB\Documents\Indexes\AbstractMultiMapIndexCreationTask;
use RavenDB\Documents\Indexes\FieldStorage;
//endregion

use RavenDB\Demo\common\DocumentStoreHolder;

//region Demo
//region Step_1
class Contacts_ByNameAndTitle extends AbstractMultiMapIndexCreationTask
{
//endregion
    public function __construct()
    {
        parent::__construct();

        //region Step_2
        $this->addMap("docs.Employees.Select(employee => new {" .
            "    ContactName = (employee.FirstName + \" \") + employee.LastName," .
            "    ContactTitle = employee.Title," .
            "    Collection = this.MetadataFor(employee)[\"@collection\"]" .
            "})");

        $this->addMap("docs.Companies.Select(company => new {" .
            "    ContactName = company.Contact.Name," .
            "    ContactTitle = company.Contact.Title," .
            "    Collection = this.MetadataFor(company)[\"@collection\"]" .
            "})");

        $this->addMap("docs.Suppliers.Select(supplier => new {" .
            "    ContactName = supplier.Contact.Name," .
            "    ContactTitle = supplier.Contact.Title," .
            "    Collection = this.MetadataFor(supplier)[\"@collection\"]" .
            "})");
        //endregion
        
        //region Step_3
        $this->store("ContactName", FieldStorage::yes());
        $this->store("ContactTitle", FieldStorage::yes());
        $this->store("Collection", FieldStorage::yes());
        //endregion
    }
}
//endregion

//region Demo
//region Step_4
class IndexEntry
{
    public ?string $ContactName = null;
    public ?string $ContactTitle = null;
    public ?string $Collection = null;

    public function getContactName(): ?string
    {
        return $this->ContactName;
    }

    public function setContactName(?string $ContactName): void
    {
        $this->ContactName = $ContactName;
    }

    public function getContactTitle(): ?string
    {
        return $this->ContactTitle;
    }

    public function setContactTitle(?string $ContactTitle): void
    {
        $this->ContactTitle = $ContactTitle;
    }

    public function getCollection(): ?string
    {
        return $this->Collection;
    }

    public function setCollection(?string $Collection): void
    {
        $this->Collection = $Collection;
    }
}
//endregion
//endregion

//region Demo
//region Step_5
class ProjectedEntry extends IndexEntry
{
    public ?string $phone = null;

    public function getPhone(): ?string
    {
        return $this->phone;
    }

    public function setPhone(?string $phone): void
    {
        $this->phone = $phone;
    }
}
//endregion
//endregion

class MultiMapIndexCustomizedFields
{
    public function __invoke(RunParams $runParams): array
    {
        $namePrefix = $runParams->getNamePrefix() ?? "Michael";
        $titlePrefix = $runParams->getTitlePrefix() ?? "Sales";

        //region Demo
        //region Step_6
        $contacts = [];

        $session = DocumentStoreHolder::getStore()->openSession();
        
        try {
            $contacts = $session->query(IndexEntry::class, Contacts_ByNameAndTitle::class)
                ->whereStartsWith("ContactName", $namePrefix)
                ->whereStartsWith("ContactTitle", $titlePrefix)
                ->selectFields(ProjectedEntry::class)
                ->toList();
                
        } finally {
            $session->close();
        }
        //endregion
        //endregion
        
        return $contacts;
    }
}
